### ISSUE TITLE
JWT claims removed from token

### ISSUE THREAD

**Issue by Maxtoid** [causefx/Organizr#1577]
###### Operating System: Windows/MacOS/Ubuntu - Yes
<hr>

##### Problem Description:
Seems that most of the user specific claims were removed from the jwt token in commit [10019633bea1e5635e14c75b06e09c1d54bf4507](https://github.com/causefx/Organizr/commit/10019633bea1e5635e14c75b06e09c1d54bf4507)

Are any of these going to be re-added at any point? Unless there's another way that I've missed, it seems like the only way to successfully authenticate with Organizr using Caddy 2 is using the caddy-auth-jwt plugin for Caddy V2, which doesn't allow you to specify unique claims. 

I've been able to get logged in / vs. not logged in authentication working in Caddy 2, but cannot limit hosts based on user group due to "**group**" being removed from the organizr token. Would be awesome if at least "group" could be re-added to the token at some point to enable jwt token at some point so Organizr auth can be fully functional with Caddy 2.

_(I've documented my exhaustive journey getting Caddy 2 working with Organizr JWT and am writing up a blog post on it as soon as I have it working with group specification, I'll share the link once it's done)_

<hr>

##### Reproduction Steps:
examine the organizr_token 

Only iss, aud, jti, iat, exp, & userID are passed - username, group, groupID, email, & image are missing. Need at least **group** for caddy 2 jwt auth to work properly.

<hr>

**Comment by strothj** [comment]
Thanks for posting this, I thought I was going crazy. I updated Organizr and I also noticed the JWT claims had disappeared. I'm also using `caddy-auth-jwt`.

I was debating rolling back the Organizr version or perhaps using the UserId fields in the meantime (less than ideal).

**Comment by causefx** [comment]
I can add back in the GroupID

**Comment by strothj** [comment]
Hmm, actually I believe we need one of the following fields in the JWT:
- roles
- role
- group
- groups
- app_metadata - authorization - roles
- realm_access - roles

https://github.com/greenpau/caddy-auth-jwt#sources-of-role-information

I didn't see a way to use a custom `roles` field. It looks like it is a future feature. If possible, it would be great to add back the `group` field.

**Comment by causefx** [comment]
i can have group return the group id.  would that work?

**Comment by strothj** [comment]
api/classes/organizr.class.php:
```php
->withClaim('group', strval($result['group_id']))// Configures a new claim, called "groupID"
```

The Caddy middleware expects the claim to be a string, otherwise it throws an error about the claim being a float. If I convert the field to a string, everything works fine. I'm not a PHP dev so that string conversion was from a Google search :-P.

The Caddyfile changes look like this:
```
		jwt {
			allow roles Admin Co-Admin
		}
```
to
```
		jwt {
			allow roles 0 1
		}
```

jwt:
```json
{
  "iss": "Organizr",
  "aud": "Organizr",
  // ...
  "group": "0",
  "userID": 1
}
```

Does this sound reasonable to you @Maxtoid @causefx ?

**Comment by Maxtoid** [comment]
Yeah I think either having **group** return group ID, or honestly just even putting group back in and returning the name of the group would work; the way the JWT plugin processes the strings, it would be easier to just allow roles **admin**, **co-admin**, or **superuser**. But if the size of the token is the concern we should be able to get the JWT token to work with the group ID as the group claim too.

**Comment by causefx** [comment]
Yeah, I was trying to cut down on the length of the token to limit the cookie sizes...  I will play with it and release it tomorrow.

**Comment by systemofapwne** [comment]
This seem to reflect the issue I face right now. I just started writing a small node script to parse the JWT in order to lift the load  on the already nice organizr auth backend. My situation is basically that nginx hammers the auth backend like crazy for certain services.
In contrast to the [wiki entry](https://docs.organizr.app/books/setup-features/page/serverauth), I noticed that the JWT only contains limited info like

``
{
  iss: 'Organizr',
  aud: 'Organizr',
  jti: '<REDACTED>',
  iat: 1613858468,
  exp: 1616277668,
  group: 'Admin',
  userID: 1
}
``
I will now code around that and query the sqlite db for the groupId a user actually is in (incl. some caching to reduce load)

I (and I guess many others) would be fine, if the groupId gets added back though.
However I can understand, that removing the groupId would make sense for the case, when a user is moved across groups. But then I do not understand why the "group" string is still available.
So in short: 
- For convenience, having the groupId at least might be cool.
- In terms of safety, having only the userid in the string and manually querying against the group makes most sense
- But the later case will cause IO load again and might be in the end slow (or even slower) than directly querying the organizr "auth backend".

I will, as stated before, opt for the userId query to the current groupId via sqlite (with some in-memory caching).

**Comment by strothj** [comment]
@systemofapwne 

If you take advantage of the JWT for authorization, there shouldn't be any load on Organizr.

The field `group` maps 1:1 to the previously provided `groupId` field, with the exception that it's a string enum. There shouldn't be any need to query Orgnaizr to determine what group a user belongs to given that the JWT provides that with the `group` field. I have this working successfully with Caddy server so I'd expect it to work similarly with Nginx.

A route on my Caddy 2 server is configured like so:
```
service.domain.tld {
	import common

	route * {
		jwt {
			import jwt-primary
			allow roles Admin Co-Admin
		}

		reverse_proxy http://service
	}
}
```

**Comment by systemofapwne** [comment]
Thank you for this very fast reply!

> If you take advantage of the JWT for authorization, there shouldn't be any load on Organizr.

Yes, this is exactly my plan

> 
> The field `group` maps 1:1 to the previously provided `groupId` field, with the exception that it's a string enum. There shouldn't be any need to query Orgnaizr to determine what group a user belongs to given that the JWT provides that with the `group` field. I have this working successfully with Caddy server so I'd expect it to work similarly with Nginx.

I am currently on v2-develop with _Organizr Version 2.1.195_.
When I B64 decode the payload with e.g. notepad++, it literally prints out to

``{"iss":"Organizr","aud":"Organizr","jti":"<REDACTED>","iat":1613858468,"exp":1616277668,"group":"Admin","userID":1}``

So the string-enum is casted to a string literal rather than a string-enum. That's why I struggle. I really would like to use the enumerated ID for my scripts. But that one seems to be lost in translation.

> 
> A route on my Caddy 2 server is configured like so:
> 
> ```
> service.domain.tld {
> 	import common
> 
> 	route * {
> 		jwt {
> 			import jwt-primary
> 			allow roles Admin Co-Admin
> 		}
> 
> 		reverse_proxy http://service
> 	}
> }
> ```

I see that you use the bare strings here. That might be OK for many use cases and as long as you do not rename those groups. However I would prefer to use the groupID via the old "/auth-0" etc nginx auth-backend against my JWT script (to keep the transition to JWT as minimal as possible). So I rather rely on groupIds. Those are however missing, since the group entry in the JWT object is of type string.

But as stated before: I will now opt for queriying the sqlite.db from time to time in my JWT node-script to have an updated user to group association in the case of, that I need to shift a user to another group.
However, the JWT group should then still be of type "string-enum" to get the groupId from it rather than plain "string"

**Comment by causefx** [comment]
Does anything else need to be done with this issue?

**Comment by strothj** [comment]
> Does anything else need to be done with this issue?

Things seem to be working fine on my end at least. I don't know about others.

**Comment by Maxtoid** [comment]
> > Does anything else need to be done with this issue?
> 
> 
> 
> Things seem to be working fine on my end at least. I don't know about others.

All good for me as well. Thanks!!

**Comment by Yann-J** [comment]
Hello there! 👋

I'm tempted to reopen this ticket since my request is so similar.

I was also affected by the group names being dropped from the JWT claims, which I'm also using for some basic group-based access control, and that was addressed properly.

However, I was also using the token content to support _proxy authentication_ in some apps (such as Calibre-web - see [here](https://github.com/janeczku/calibre-web/wiki/Setup-Reverse-Proxy) for info on how it works).

Basically some apps like calibre-web can be configured to automatically trust the authentication performed by the reverse proxy (Caddy in my case) and read the user name from a request header filled by the proxy (which of course is also stripped from the original user's requests by the proxy to prevent impersonations). The users need to be created upfront, but the authentication is then completely transparent.

Is it possible to also restore the usernames in the JWT cookie?

Many thanks!

**Comment by causefx** [comment]
completed.

**Comment by Yann-J** [comment]
Wonderful, thanks! 😍
